### PR TITLE
Create sample conditions (2/3)

### LINKED ISSUE #691: Create `sample_conditions()` method
### Problem Description

Let's make sampling more user friendly. We can create multiple methods for different user needs.

A new `sample_conditions` method can address conditional sampling needs.

### Expected behavior

Parameters:
- (required) `conditions` : A list of `Condition` objects (see #689) -- this specifies `num_rows`
- `max_tries`: renamed from existing `max_retries` param (default: 100)
- `batch_size_per_try`: Number of rows to sample per try (default: 10x requested num)
- `randomize_samples` will determine whether or not there should be a fixed seed (default: True)

```python
# works with any tabular model
from sdv.tabular import CTGAN
model = CTGAN()
model.fit(data)

# see Issue #689 for Condition object details
from sdv.tabular.sampling import Conditions
female_users = Condition(column_values={'sex': 'F', 'active_user': True}, num_rows=50)
inactive_users = Condition(column_values={'active_user': False}, num_rows=100)
conditions = [female_users, inactive_users]

# pass in list of conditions
model.sample_conditions(conditions, max_retries=200, randomize_sample=False)
```

## Error Handling
Running out of tries
```python
# Always gracefully reject sample (ie return any rows that are sampled)
>>> synthetic_data = model.sample_conditions(conditions)
Warning: Only able to sample 75 rows for the given conditions. To sample more rows, try increasing max_tries
(currently: 100) or increasing batch_size_per_try (currently: 10000). Note that increasing these values will also
increase the sampling time.

# Error if we weren't able to sample any rows
>>> synthetic_data = model.sample_conditions(conditions)
Error: Unable to sample any rows for the given conditions. Try increasing max_tries
(currently: 100) or increasing batch_size_per_try (currently: 10000). Note that increasing these values will also
increase the sampling time.
```

Checking for invalid input
```python
# Unexpected column name
>>> c = Condition(column_values={'New_Column': 42})
>>> synthetic_data = model.sample_conditions([c])
Error: Unexpected column name 'New_Column'. Use a column name that was present in the original data.
 
# Unexpected categorical column value
>>> c = Condition(column_values={'sex': 'NEW_VALUE'}, num_rows=1)
Error: Unexpected value 'NEW_VALUE' in column 'New_Column'. Use a value that was present in the original data.
```

### PR DESCRIPTION
**Pull request by katxiao**
Resolves #691

### CODE REVIEW

File: sdv/tabular/base.py [hunk 1]
```diff
@@ -320,75 +319,69 @@ def _sample_batch(self, num_rows=None, max_retries=100, max_rows_multiplier=10,
...
-                num_to_sample, conditions, transformed_conditions, float_rtol, sampled
+                batch_size_per_try, conditions, transformed_conditions, float_rtol, sampled,
             )
 
             counter += 1
 
         return sampled.head(min(len(sampled), num_rows))
 
-    def _make_conditions_df(self, conditions, num_rows):
+    def _make_conditions_df(self, conditions):
```
Comment (amontanez24) [inline sdv/tabular/base.py:344]: should the name change since this now returns a list?

File: sdv/tabular/base.py [hunk 2]
```diff
@@ -439,6 +435,134 @@ def sample(self, num_rows, randomize_samples=True):
...
+        """
+        for column in conditions.columns:
+            if column not in self._metadata.get_fields():
+                raise ValueError(f'Error: Unexpected column name `{column}`. '
+                                 f'Use a column name that was present in the original data.')
+
+        try:
+            transformed_conditions = self._metadata.transform(conditions, on_missing_column='drop')
+        except ConstraintsNotMetError as cnme:
+            cnme.message = 'Passed conditions are not valid for the given constraints'
```
Comment (amontanez24) [inline sdv/tabular/base.py:472]: minor: can we say provided conditions instead of passed

File: tests/unit/tabular/test_base.py [hunk 3]
```diff
@@ -19,6 +21,57 @@
 
 class TestBaseTabularModel:
 
+    def test_sample_no_transformed_columns(self):
```
Comment (amontanez24) [inline tests/unit/tabular/test_base.py:24]: this should be `test__sample_with_conditions...` right?

File: tests/unit/tabular/test_base.py [hunk 4]
```diff
@@ -179,6 +278,33 @@ def test__init__passes_correct_parameters(metadata_mock):
     metadata_mock.assert_has_calls(expected_calls, any_order=True)
 
 
+@pytest.mark.parametrize('model', MODELS)
+def test_conditional_sampling_graceful_reject_sampling(model):
```
Comment (amontanez24) [inline tests/unit/tabular/test_base.py:282]: now that things are cleaned up, can we rename this to `test_sample_conditions`?

File: tests/unit/tabular/test_base.py [hunk 5]
```diff
@@ -224,6 +350,130 @@ def test__sample_rows_previous_rows_appended_correctly():
     pd.testing.assert_frame_equal(sampled, expected)
 
 
+def test_sample_empty_transformed_conditions():
```
Comment (amontanez24) [inline tests/unit/tabular/test_base.py:353]: Can we just rename any tests that use the new methods to have that method name in them?

File: sdv/tabular/base.py (hunk 1 above)
Comment (katxiao) [reply sdv/tabular/base.py]: Sure, done!

File: sdv/tabular/base.py [hunk 6]
```diff
@@ -320,75 +319,69 @@ def _sample_batch(self, num_rows=None, max_retries=100, max_rows_multiplier=10,
...
-        elif num_rows is not None and len(conditions) != num_rows:
-            raise ValueError(
-                'If `conditions` is a `DataFrame`, `num_rows` must be `None` or match its lenght.')
+        condition_dataframes = defaultdict(list)
+        for condition in conditions:
+            column_values = condition.get_column_values()
+            condition_dataframes[tuple(column_values.keys())].append(
+                pd.DataFrame(
+                    condition.get_column_values(),
+                    index=range(condition.get_num_rows())
```
Comment (fealho) [inline sdv/tabular/base.py:363]: Doesn't this get set automatically? Do we really need this line?

File: sdv/tabular/base.py [hunk 7]
```diff
@@ -431,6 +427,134 @@ def sample(self, num_rows, randomize_samples=True):
...
+            ValueError:
+                If any of the following happens:
+                    * any of the conditions' columns are not valid.
+                    * `graceful_reject_sampling` is `False` and not enough valid rows could be
+                      sampled within `max_tries` trials.
+                    * no rows could be generated.
+        """
+        for column in conditions.columns:
+            if column not in self._metadata.get_fields():
+                raise ValueError(f'Error: Unexpected column name `{column}`. '
```
[the remark this answers is not in the document]
Comment (katxiao) [reply sdv/tabular/base.py:458]: This is fixed in a later PR

File: sdv/tabular/base.py [hunk 8]
```diff
@@ -300,15 +302,12 @@ def _sample_batch(self, num_rows=None, max_retries=100, max_rows_multiplier=10,
                 Number of rows to sample. If not given the model
                 will generate as many rows as there were in the
                 data passed to the ``fit`` method.
-            max_retries (int):
-                Number of times to retry sampling discarded rows.
+            max_tries (int):
```
[the remark this answers is not in the document]
Comment (katxiao) [reply sdv/tabular/base.py:305]: With the changes in [this PR](https://github.com/sdv-dev/SDV/pull/709/files), I believe that issue will be fixed.

File: sdv/tabular/base.py (hunk 6 above)
Comment (katxiao) [reply sdv/tabular/base.py:363]: A condition could be defined as
```
condition = Condition({'a': 1}, num_rows=5)
```
which means that the user wants 5 rows with the condition that column `a` is 1, so in the dataframe we want `{'a': 1}` to repeat 5 times. Setting the index to range(5) does that.

File: sdv/tabular/base.py [hunk 9]
```diff
@@ -399,12 +392,15 @@ def _conditionally_sample_rows(self, dataframe, max_retries, max_rows_multiplier
...
 
             elif not graceful_reject_sampling:
-                error = f'Could not get enough valid rows within {max_retries} trials.'
+                error = f'Could not get enough valid rows within {max_tries} trials.'
                 raise ValueError(error)
 
             else:
-                warn(f'Only {len(sampled_rows)} rows could '
-                     f'be sampled within {max_retries} trials.')
+                warn(f'Warning: Only able to sample {len(sampled_rows)} rows for the given '
```
[the remark this answers is not in the document]
Comment (katxiao) [reply sdv/tabular/base.py:399]: Not sure I understand this one - I'm importing `warn` from `warnings`, so it's essentially the same thing as calling `warnings.warn`.

[the remark this answers is not in the document]
Comment (fealho) [reply sdv/tabular/base.py]: The behavior is the same, but I think our policy is to be explicit whenever possible, like the following:

```python3
import warnings
warnings.warn('...')
```

This doesn't matter much though, just something to keep in mind.

**Review by fealho — APPROVED** [review]
